### ISSUE TITLE
Replace OperatorInterface.apply_adjoint by OperatorInterface.apply_transpose

### ISSUE THREAD

**Issue by sdrave** [pymor/pymor#332]
For a linear operator O: V -> W between Banach spaces, the transposed operator O^T: W' -> V' is the linear operator which maps a linear functional f on W to the linear functional f o O on V. Given bases of V and W (finite dimensional), the matrix of O^T w.r.t. to the associated dual bases is simply the transpose of the matrix of O w.r.t. those bases.

In pyMOR, we implicitly always assume that all vectors are vectors w.r.t. some (e.g. finite element) basis, and when we use V.dot(U) to implement functionals, we implicitly assume the functional to be given w.r.t. the dual basis. Thus, it is natural to have an `apply_transpose` method in `OperatorInterface` which for a NumpyMatrixOperator is given by left-multiplication with the vector.

Currently, we have in pyMOR the notion of an `AdjointOperator`, e.g. the Operator O^*: W -> V, such that (w, Ov)_W = (O^*w, v)_V., where (., .)_V, (., .)_W denote the inner products on the *Hilbert spaces* V and W.

Checking the definitions, it is straightforward to see that O^* = R_V^{-1} o O^T o R_W, where R_W: W -> W' is the Risz isomorphism mapping w to the functional (w, .). In coordinates, R_W is simply multiplication with the W-inner product matrix, whereas R_V^{-1} is the multiplication with the inverse of the V-inner product matrix. These product matrices are given in pyMOR as the `source_product` and `range_product` arguments of `apply_adjoint`.

Thus:
- O^* makes only sense when V and W are Hilbert spaces.
- 'apply_adjoint' has a simple default implementation based on O^T (i.e. the case where `source_product` and `range_product` are `None`).
- O^T always makes sense, even when V and W are not Hilbert spaces by nature.

Right now, I really do not know any use-case where `apply_adjoint` is used with non-trivial `source_product` or `range_product`, and I would doubt that many users understand their meaning. On the other hand, having these arguments is in many cases responsibly for most lines of code in an `apply_adjoint` or `apply_inverse_ajoint` implementation (e.g. see `NumpyMatrixOperator`).  Therefore, I propose the following:

- Rename `apply_adjoint` to `apply_transpose` and remove `source_product` and `range_product` arguments.
- Keep `AdjointOperator` in `pymor.operators.constructions` in case someone really needs the adjoint operator with non-trivial products.
- Rename `apply_inverse_adjoint` to `apply_inverse_transpose` and remove `source_product` and `range_product` arguments.
- Rename `inverse_adjoint` solver option to `inverse_transpose`.

Downside of this proposal would be, that it would be impossible to exploit a specialized implementation of the adjoint operator (for non-trivial products) in generic code which now typically would call `AdjointOperator(op, source_product=sp, range_product=rp).apply(V)`.

Any objections, @ftalbrecht, @pmli, @andreasbuhr? Otherwise I will start implementing this soon.

**Comment by pmli** [comment]
No objections from my side.

**Comment by ftalbrecht** [comment]
Sounds good to me!

**Comment by sdrave** [comment]
Of course, it would also be possible, to add `apply_transpose` and default implement `apply_adjoint` via `apply_transpose` ...

**Comment by pmli** [comment]
In this case, I think `apply_adjoint` should behave as conjugate transpose (so far, `NumpyMatrixOperator.apply_adjoint` didn't conjugate the matrix).

**Comment by sdrave** [comment]
Good point! How to achieve that would largely depend on how we decide to treat antilinear operators in pyMOR, though.

**Comment by pmli** [comment]
Adjoint (or conjugate transpose) operator is linear, not antilinear, as far as I understand.

**Comment by sdrave** [comment]
Sure. But for a default implementation you would need the Riesz isomorphism, which is antilinear.

**Comment by andreasbuhr** [comment]
I do rely on the apply_adjoint method with non trivial products. I use it to calculate an operator-SVD. But I could also just use AdjointOperator from "constructions". 
So I agree this is a good idea.
